src/Store/VimStoreConnector.re

```diff
@@ -425,6 +425,118 @@ let start =
   let lastCompletionMeet = ref(None);
   let isCompleting = ref(false);
 
+  let contextFromState: State.t => Vim.Context.t =
+    state => {
+      state
+      |> Selectors.getActiveEditorGroup
+      |> Selectors.getActiveEditor
+      |> Option.map((editor: Editor.t) => {
+           let {bufferId, cursors, _}: Editor.t = editor;
+           let primaryCursor = Editor.getPrimaryCursor(editor);
+
```

Comment:
This looks so much better! It's much easier (assuming no bugs on the libvim side of things) to find out exactly what state libvim is in for an operation, and to see exactly what's being "synced" with this.

One possible improvement could be to make it even more granular, like the context keys, where every "piece" is calculated in isolation based on the state. This function is already over one hundred lines long, and so to fully understand a calculation at the bottom, you really have to take every one of those 100+ lines into account to figure out what each component of the calculation actually is.

Also, having a 100+ line anonymous function in a pipe chain isn't great, for much the same reason.

Still a massive step in the right direction though!

Reply:
FYI, I pulled apart this method and split it up in a `VimContext` module in `Oni_Model`

Reply:
Awesome!

> I pulled apart this method

Very nitpicky, but pet peeve and such, you know... It's a function, not a method. "method" is an OOP-term for a function that's attached to an object.  Outside of OOP there are no methods, just functions. Or procedures, if you're very old and imperatively inclined ;)

Reply:
Sorry, leaked some OOP in, thanks for the correction ;)